store/sqlstore/atomic_store.go

```diff
@@ -0,0 +1,221 @@
+				}
+			}
+		}
+	} else if ps.DriverName() == model.DATABASE_DRIVER_MYSQL {
+		sql := `INSERT INTO AtomicKeyValueStore (AKey, AValue, ExpireAt, UpdateAt)
+			VALUES(:Key, :Value, :ExpireAt, :UpdateAt)
+			ON DUPLICATE KEY UPDATE AValue = :Value, ExpireAt = :ExpireAt, UpdateAt = :UpdateAt`
+		if _, err := ps.GetMaster().Exec(sql, map[string]interface{}{"Key": kv.Key, "Value": kv.Value, "ExpireAt": kv.ExpireAt, "UpdateAt": kv.UpdateAt}); err != nil {
+			return nil, model.NewAppError("SqlAtomicStore.SaveOrUpdate", "store.sql_atomic_store.save.app_error", nil, err.Error(), http.StatusInternalServerError)
+		}
+	}
+
+	return kv, nil
+}
+
+func (ps SqlAtomicStore) CompareAndSet(kv *model.AtomicKeyValue, oldValue []byte) (bool, *model.AppError) {
+	kv.PreSave()
+	if err := kv.IsValid(); err != nil {
+		return false, err
+	}
+
+	if kv.Value == nil {
+		// Setting a key to nil is the same as removing it
+		return ps.CompareAndDelete(kv.Key, oldValue)
+	}
+
+	if oldValue == nil {
+		// Insert if oldValue is nil
+		if err := ps.GetMaster().Insert(kv); err != nil {
+			// If the error is from unique constraints violation, it's the result of a
+			// race condition, return false and no error. Otherwise we have a real error and
+			// need to return it.
+			if IsUniqueConstraintError(err, []string{"PRIMARY", "Key", "AKey"}) {
+				return false, nil
+			} else {
+				return false, model.NewAppError("SqlAtomicStore.CompareAndSet", "store.sql_atomic_store.save.app_error", nil, err.Error(), http.StatusInternalServerError)
+			}
+		}
+	} else {
+		// Update if oldValue is not nil
+		updateResult, err := ps.GetMaster().Exec(
+			`UPDATE AtomicKeyValueStore SET AValue = :New, ExpireAt = :ExpireAt, UpdateAt = :UpdateAt WHERE AKey = :Key AND AValue = :Old`,
+			map[string]interface{}{
+				"Key":      kv.Key,
+				"Old":      oldValue,
+				"New":      kv.Value,
+				"ExpireAt": kv.ExpireAt,
+				"UpdateAt": kv.UpdateAt,
+			},
+		)
+		if err != nil {
+			return false, model.NewAppError("SqlAtomicStore.CompareAndSet", "store.sql_atomic_store.save.app_error", nil, err.Error(), http.StatusInternalServerError)
+		}
+
+		if rowsAffected, err := updateResult.RowsAffected(); err != nil {
+			// Failed to update
+			return false, model.NewAppError("SqlAtomicStore.CompareAndSet", "store.sql_atomic_store.save.app_error", nil, err.Error(), http.StatusInternalServerError)
+		} else if rowsAffected == 0 {
+			if ps.DriverName() == model.DATABASE_DRIVER_MYSQL && bytes.Equal(oldValue, kv.Value) {
+				// ROW_COUNT on MySQL is zero even if the row existed but no changes to the row were required.
+				// Check if the row exists with the required value to distinguish this case. Strictly speaking,
+				// this isn't a good use of CompareAndSet anyway, since there's no corresponding guarantee of
+				// atomicity. Nevertheless, let's return results consistent with Postgres and with what might
+				// be expected in this case.
+				count, err := ps.GetReplica().SelectInt(
```

Comment:
I am wondering whether we should do the select against master itself ? Because we just made a query against master and expect the results of it to be immediately available in a replica ? Will there be issues if there is a lag in master-replica synchronization ?

Reply:
fixed :)